crates/red_knot/src/semantic/types.rs

```diff
@@ -520,6 +529,29 @@ pub struct UnionTypeId {
     union_id: ModuleUnionTypeId,
 }
 
+impl UnionTypeId {
+    pub fn elements(self, type_store: &TypeStore) -> ElementsIterator {
+        let union = type_store.get_union(self);
+        ElementsIterator {
+            elements: union.elements.iter().rev().copied().collect(),
+        }
+    }
+}
+
+pub struct ElementsIterator {
+    elements: Vec<Type>,
+}
+
+impl Iterator for ElementsIterator {
+    type Item = Type;
```

Comment:
It's a bit weird to me that we have to call `.collect()` here in order to create the iterator instance. Normally I wouldn't expect a method returning an iterator to allocate anything, but here there is an allocation under the hood.

If we're already collecting, why don't we just return `Vec<Type>` from `.elements()`? You can do more with a `Vec` than you can with an iterator. Or do you not want us to start relying on that API?

Reply:
Yeah, you're right, this should just return an owned vec. I might not even need it at all if I do the imperative version above.